accounts/abi/topics.go

```diff
@@ -37,71 +37,99 @@ import (
 	"github.com/ethereum/go-ethereum/crypto"
 )
 
+// packTopic attempts to pack rule into the corresponding hash value for a log's topic
+// according to the Solidity documentation:
+// https://docs.soliditylang.org/en/v0.8.17/abi-spec.html#indexed-event-encoding.
+func packTopic(rule interface{}) (common.Hash, error) {
+	var topic common.Hash
+
+	// Try to generate the topic based on simple types
+	switch rule := rule.(type) {
+	case common.Hash:
+		copy(topic[:], rule[:])
+	case common.Address:
+		copy(topic[common.HashLength-common.AddressLength:], rule[:])
+	case *big.Int:
+		blob := rule.Bytes()
+		copy(topic[common.HashLength-len(blob):], blob)
+	case bool:
+		if rule {
+			topic[common.HashLength-1] = 1
+		}
+	case int8:
+		copy(topic[:], genIntType(int64(rule), 1))
+	case int16:
+		copy(topic[:], genIntType(int64(rule), 2))
+	case int32:
+		copy(topic[:], genIntType(int64(rule), 4))
+	case int64:
+		copy(topic[:], genIntType(rule, 8))
+	case uint8:
+		blob := new(big.Int).SetUint64(uint64(rule)).Bytes()
+		copy(topic[common.HashLength-len(blob):], blob)
+	case uint16:
+		blob := new(big.Int).SetUint64(uint64(rule)).Bytes()
+		copy(topic[common.HashLength-len(blob):], blob)
+	case uint32:
+		blob := new(big.Int).SetUint64(uint64(rule)).Bytes()
+		copy(topic[common.HashLength-len(blob):], blob)
+	case uint64:
+		blob := new(big.Int).SetUint64(rule).Bytes()
+		copy(topic[common.HashLength-len(blob):], blob)
+	case string:
+		hash := crypto.Keccak256Hash([]byte(rule))
+		copy(topic[:], hash[:])
+	case []byte:
+		hash := crypto.Keccak256Hash(rule)
+		copy(topic[:], hash[:])
+
+	default:
+		// todo(rjl493456442) according solidity documentation, indexed event
+		// parameters that are not value types i.e. arrays and structs are not
+		// stored directly but instead a keccak256-hash of an encoding is stored.
+		//
+		// We only convert strings and bytes to hash, still need to deal with
+		// array(both fixed-size and dynamic-size) and struct.
+
+		// Attempt to generate the topic from funky types
+		val := reflect.ValueOf(rule)
+		switch {
+		// static byte array
+		case val.Kind() == reflect.Array && reflect.TypeOf(rule).Elem().Kind() == reflect.Uint8:
+			reflect.Copy(reflect.ValueOf(topic[:val.Len()]), val)
+		default:
+			return common.Hash{}, fmt.Errorf("unsupported indexed type: %T", rule)
+		}
+	}
+	return topic, nil
+}
+
+// PackTopics attempts to pack the array of filters into an array of corresponding topics
+// according to the Solidity documentation.
+// Note: PackTopics does not support array(both fixed size and dynamic-size) and struct types.
+func PackTopics(filter []interface{}) ([]common.Hash, error) {
```

Comment:
We should move this into the ABI package and have a single function to pack both the topics and event input data.

Reply:
not sure I follow this comment.

Reply:
Updated the PR with what I was thinking.

Logs include both topics `[]common.Hash` and event data, which is more similar to the way function arguments are encoded and `PackTopics` only converts arguments to the corresponding topic hash based off of this.

Now there is `PackEvent` which will take in all of the arguments both indexed and non-indexed and return both the resulting topics and event data.